tools/asset-sync/assets.Tests.ps1

```diff
@@ -0,0 +1,430 @@
+
+      Push-AssetsRepo-Update -Config $Config
+
+      try {
+        Push-Location $assetRepoFolder
+        $repoBranchSHA = git rev-parse origin/$($config.AssetsRepoBranch) --quiet
+      }
+      finally {
+        Pop-Location
+      }
+
+      # re-parse from the on-disk json
+      $configReparsed = Resolve-AssetsJson (Join-Path $testFolder "sdk" "tables" "azure-data-tables")
+
+      # re-parsed config from disk should
+      #  -> match the updated sha
+      #  -> match the SHA we get back from the repo
+      $repoBranchSHA | Should -Be $config.SHA
+      $configReparsed.SHA | Should -Be $config.SHA
+    }
+    
+    It "Should push a clean new commmit to the target branch." {
+      $sourceBranch = "scenario_clean_push"
+      $recordingJson = [PSCustomObject]@{
+        AssetsRepo = "Azure/azure-sdk-assets-integration"
+        AssetsRepoPrefixPath = "python/recordings/"
+        AssetsRepoId = ""
+        AssetsRepoBranch = "$sourceBranch"
+        SHA = "e4a4949a2b6cc2ff75afd0fe0d97cbcabf7b67b7"
+      }
+
+      $files = @(
+        "sdk/tables/azure-data-tables/tests/recordings/test_retry.pyTestStorageRetrytest_retry_on_timeout.json",
+        "sdk/tables/azure-data-tables/tests/recordings/test_retry.pyTestStorageRetrytest_retry_on_server_error.json",
+        "sdk/tables/assets.json"
+      )
+
+      $testFolder = Describe-TestFolder -AssetsJsonContent $recordingJson -Files $files -IntegrationBranch $recordingJson.AssetsRepoBranch
+      $config = Resolve-AssetsJson (Join-Path $testFolder "sdk" "tables")
+      $config.AssetsRepoBranch | Should -not -Be $sourceBranch
+      
+      Initialize-AssetsRepo -Config $config
+      $assetRepoFolder = Resolve-AssetRepo-Location -Config $config
+
+      foreach($file in $files){
+        $sourcePath = Join-Path $testFolder $file 
+        $targetPath = Join-Path $assetRepoFolder $config.AssetsRepoPrefixPath $file
+
+        $targetFolder = Split-Path $targetPath
+        if(-not (Test-Path $targetFolder)){
+          mkdir -p $targetFolder
+        }
+
+        Copy-Item  -Force -Path $sourcePath -Destination $targetPath
+      }
+      $config.AssetsRepoBranch | Should -not -Be $sourceBranch
+
+      Push-AssetsRepo-Update -Config $Config
+
+      try {
+        Push-Location $assetRepoFolder
+        $repoBranchSHA = git rev-parse origin/$($config.AssetsRepoBranch) --quiet
+      }
+      finally {
+        Pop-Location
+      }
+
+      # re-parse from the on-disk json
+      $configReparsed = Resolve-AssetsJson (Join-Path $testFolder "sdk" "tables" "azure-data-tables")
+
+      # re-parsed config from disk should
+      #  -> match the updated sha
+      #  -> match the SHA we get back from the repo
+      $repoBranchSHA | Should -Be $config.SHA
+      $configReparsed.SHA | Should -Be $config.SHA
+    }
+
+    It "Should push a new commit on top of an existing one. Doing a pre-fetch if necessary." {
+      $sourceBranch = "scenario_conflict_push"
```

Comment:
If this test is pushing changes, how will that work if multiple people are running the test? Perhaps the branch name should have a dynamic suffix and get cleaned up at the end?

Reply:
Or should test setup require a fork?

Reply:
This is what it does. Check `Describe-Test-Folder`.

We actually grab your target branch. If it already exists, we push nothing, we just say

> your new target branch is `test_<branchname>_<guid suffix>`

and update your assets.json with it before writing all our files into the fake language repo construct.

If the branch _does_ exist upstream, we deliberately duplicate and push it upstream, then update local references to point at it.